File: 0101-0200/0173.Binary_Search_Tree_Iterator.java
//17ms







/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {
    Stack<TreeNode> st = new Stack<>(); // stack to store nodes
    
    public BSTIterator(TreeNode root) {
        // initialize stack with the left chain of root
        pushLeftChain(root);
    }
    
    public int next() {
        // get the next smallest node
        TreeNode node = st.pop();
        
        // if node has a right child, push its left chain
        if (node.right != null) {
            pushLeftChain(node.right);
        }
        
        return node.val; // return value
    }
    
    public boolean hasNext() {
        // still nodes left if stack is not empty
        return !st.isEmpty();
    }
    
    private void pushLeftChain(TreeNode node) {
        // push all left children into stack
        while (node != null) {
            st.push(node);
            node = node.left;
        }
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */







//16ms







/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {
    TreeNode curr;
    public BSTIterator(TreeNode root) {
        curr = root;
    }
    
    public int next() {
        while(curr != null){
            if(curr.left == null){
                TreeNode node = curr;
                curr = curr.right;
                return node.val;
            } else {
                TreeNode node = curr.left;
                while(node.right != null && node.right != curr){
                    node = node.right;
                }
                if(node.right == null){
                    node.right = curr;
                    curr = curr.left;
                } else {
                    TreeNode node1 = curr;
                    node.right = null;
                    curr = curr.right;
                    return node1.val;
                }
            }
        }
        return -1;
    }
    
    public boolean hasNext() {
        return curr != null;
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: 0101-0200/0173.Binary_Search_Tree_Iterator.rs
//3ms






// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::{
    cell::RefCell,
    rc::Rc,
};
type TNode = Option<Rc<RefCell<TreeNode>>>;

struct BSTIterator {
    stack: Vec<i32>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl BSTIterator {
    fn new(root: Option<Rc<RefCell<TreeNode>>>) -> Self {
        fn push(root: &TNode, stack: &mut Vec<i32>) {
            match root {
                Some(root) => {
                    let node = root.as_ref().borrow();
                    push(&node.right, stack);
                    stack.push(node.val);
                    push(&node.left, stack);
                }
                None => {}
            }
        }
        let mut stack = Vec::new();
        push(&root, &mut stack);
        Self { stack }
    }

    fn next(&mut self) -> i32 {
        self.stack.pop().unwrap()
    }

    fn has_next(&self) -> bool {
        !self.stack.is_empty()
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * let obj = BSTIterator::new(root);
 * let ret_1: i32 = obj.next();
 * let ret_2: bool = obj.has_next();
 */






//7ms






// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
struct BSTIterator {
    traversal: Vec<i32>,
}

use std::rc::Rc;
use std::cell::RefCell;
/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl BSTIterator {

    fn new(root: Option<Rc<RefCell<TreeNode>>>) -> Self {
        let mut traversal = Vec::new();
        build_inorder_traversal(root, &mut traversal);
        BSTIterator { traversal }
    }
    
    fn next(&mut self) -> i32 {
        self.traversal.pop().unwrap()
    }
    
    fn has_next(&self) -> bool {
        return !self.traversal.is_empty();
    }

}

pub fn build_inorder_traversal(current: Option<Rc<RefCell<TreeNode>>>, traversal: &mut Vec<i32>) {
    if let Some(node) = current {
        let node_ref = node.borrow();
        //invert for the stack order
        build_inorder_traversal(node_ref.right.clone(), traversal);
        traversal.push(node_ref.val);
        build_inorder_traversal(node_ref.left.clone(), traversal);
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * let obj = BSTIterator::new(root);
 * let ret_1: i32 = obj.next();
 * let ret_2: bool = obj.has_next();
 */




//3ms





// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }

    use std::cell::RefCell;
    use std::rc::Rc;
struct BSTIterator {
        values: Vec<i32>,
        counter: usize,
        end: usize,
    }

/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */

    impl BSTIterator {
        pub fn new(root: Option<Rc<RefCell<TreeNode>>>) -> Self {
            let mut stack: Vec<Rc<RefCell<TreeNode>>> = Vec::new();
            let mut acc: Vec<i32> = Vec::new();
            let mut curr = root;

            while curr.is_some() || !stack.is_empty() {
                while let Some(ref node) = curr {
                    stack.push(node.clone());
                    curr = node.clone().borrow().left.clone();
                }

                let node = stack.pop();
                acc.push(node.as_ref().unwrap().borrow().val);
                curr = node.unwrap().borrow().right.clone();
            }
            let l = acc.len();

            Self {
                values: acc,
                counter: 0,
                end: l,
            }
        }

        fn next(&mut self) -> i32 {
            self.counter += 1;
            self.values[self.counter as usize - 1]
        }

        fn has_next(&self) -> bool {
            self.counter < self.end
        }
    }
/**
 * Your BSTIterator object will be instantiated and called as such:
 * let obj = BSTIterator::new(root);
 * let ret_1: i32 = obj.next();
 * let ret_2: bool = obj.has_next();
 */
